--- linux-fusion-8.1.1/linux/drivers/char/fusion/fusiondev.c	2009-03-31 16:04:32.000000000 +0300
+++ linux-fusion-8.1.1.new/linux/drivers/char/fusion/fusiondev.c	2010-01-29 00:11:53.000000000 +0200
@@ -35,6 +35,10 @@
 #include <asm/io.h>
 #include <asm/uaccess.h>
 
+#if LINUX_VERSION_CODE > KERNEL_VERSION(2, 6, 31)
+#include <linux/sched.h>
+#endif
+
 #if LINUX_VERSION_CODE >= KERNEL_VERSION(2, 6, 2)
 #include <linux/device.h>
 #endif
--- linux-fusion-8.1.1/linux/drivers/char/fusion/fusionee.c	2009-03-31 16:04:32.000000000 +0300
+++ linux-fusion-8.1.1.new/linux/drivers/char/fusion/fusionee.c	2010-01-29 00:16:23.000000000 +0200
@@ -563,7 +563,7 @@
 			if (f != fusionee && (!target || target == f->id)) {
 				struct task_struct *p;
 
-#ifdef rcu_read_lock
+#if defined(rcu_read_lock) || LINUX_VERSION_CODE > KERNEL_VERSION(2, 6, 31)
 				rcu_read_lock();
 #else
 				read_lock(&tasklist_lock);
@@ -583,7 +583,7 @@
 					}
 				}
 
-#ifdef rcu_read_unlock
+#if defined(rcu_read_unlock) || LINUX_VERSION_CODE > KERNEL_VERSION(2, 6, 31)
 				rcu_read_unlock();
 #else
 				read_unlock(&tasklist_lock);
